lib/vrl/compiler/src/compiler.rs

```diff
@@ -240,6 +222,54 @@ impl<'a> Compiler<'a> {
         Some(Group::new(expr))
     }
 
+    fn compile_root_exprs(
+        &mut self,
+        nodes: impl IntoIterator<Item = Node<ast::RootExpr>>,
+        external: &mut ExternalEnv,
+    ) -> Vec<Expr> {
+        let mut node_exprs = vec![];
+
+        // After a terminating expression, the state is stored, but the remaining expressions are checked.
+        let mut terminated_state = None;
+
+        for root_expr in nodes {
+            match root_expr.into_inner() {
+                RootExpr::Expr(node_expr) => {
+                    self.fallible_expression_error = None;
+
+                    if let Some(expr) = self.compile_expr(node_expr, external) {
+                        if let Some(error) = self.fallible_expression_error.take() {
+                            self.diagnostics.push(error);
+                        }
+
+                        if terminated_state.is_none() {
+                            let type_def = expr.type_def((&self.local, external));
+                            node_exprs.push(expr);
+                            // an expression that has the "never" type is a terminating expression
+                            if type_def.is_never() {
+                                terminated_state =
+                                    Some((self.local.clone(), external.target().cloned()))
+                            }
+                        }
+                    }
+                }
+                RootExpr::Error(err) => self.handle_parser_error(err),
+            }
+        }
+
+        if let Some((local, details)) = terminated_state {
+            self.local = local;
+            if let Some(details) = details {
+                external.update_target(details);
+            }
+        }
+
+        if node_exprs.is_empty() {
+            node_exprs.push(Expr::Noop(Noop));
+        }
```

Comment:
What's the reason for wanting to insert a noop expression here? As far as I know, there's no need to do this, even if the list of expressions is empty.